modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts

```diff
@@ -0,0 +1,173 @@
+
+  if (!extensions) {
+    return [];
+  }
+
+  switch (true) {
+    case Boolean(extensions.EXT_feature_metadata): {
+      return handleExtFeatureMetadataExtension(attributes, extensions.EXT_feature_metadata, images);
+    }
+    case Boolean(extensions.EXT_mesh_features): {
+      // TODO Add support for EXT_mesh_features extension.
+      console.warn('EXT_mesh_features extension is not supported yet');
+      return [];
+    }
+    default: {
+      console.warn("Can't handle unsupported batchIds extensions");
+      return [];
+    }
+  }
+}
+
+/**
+ * Get batchIds from EXT_feature_metadata extension.
+ * Docs - https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_feature_metadata
+ * @param attributes
+ * @param extFeatureMetadata
+ * @param textures
+ */
+function handleExtFeatureMetadataExtension(
+  attributes: {
+    [key: string]: GLTFAccessorPostprocessed;
+  },
+  extFeatureMetadata: ExtFeatureMetadata,
+  images: Image[]
+): number[] {
+  // Take only first extension object to get batchIds attribute name.
+  const featureIdAttribute =
+    extFeatureMetadata?.featureIdAttributes && extFeatureMetadata?.featureIdAttributes[0];
+
+  if (featureIdAttribute?.featureIds?.attribute) {
+    const batchIdsAttribute = attributes[featureIdAttribute.featureIds.attribute];
+    return batchIdsAttribute.value;
+  }
+
+  if (
+    featureIdAttribute?.featureIds?.hasOwnProperty('constant') &&
+    featureIdAttribute?.featureIds?.hasOwnProperty('divisor')
+  ) {
+    const featuresCount = attributes?.POSITIONS?.value.length / 3 || 0;
+    return generateImplicitFeatureIds(
+      featuresCount,
+      featureIdAttribute.featureIds.constant,
+      featureIdAttribute.featureIds.divisor
+    );
+  }
+
+  // Take only first extension object to get batchIds attribute name.
+  const featureIdTexture =
+    extFeatureMetadata?.featureIdTextures && extFeatureMetadata?.featureIdTextures[0];
+
+  if (featureIdTexture) {
+    const textureCoordinates = attributes.TEXCOORD_0.value;
+    return generateBatchIdsFromTexture(featureIdTexture, textureCoordinates, images);
+  }
+
+  return [];
+}
+
+/**
+ * Generates implicit feature ids
+ * Spec - https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_feature_metadata#implicit-feature-ids
+ * @param featuresCount
+ * @param constant
+ * @param devisor
+ */
+function generateImplicitFeatureIds(
+  featuresCount: number,
+  constant: number = 0,
+  divisor: number = 0
+): number[] {
+  let featureIds: number[] = [];
+
+  if (divisor > 0) {
+    let currentValue = constant;
+    let devisorCounter = divisor;
+
+    for (let index = 0; index < featuresCount; index++) {
+      featureIds.push(currentValue);
+
+      devisorCounter -= 1;
+
+      if (devisorCounter === 0) {
+        currentValue++;
+        devisorCounter = divisor;
+      }
+    }
+  } else {
+    featureIds = Array<number>(featuresCount).fill(constant, 0, featuresCount);
```

Comment:
Nit: Not sure how long these are, but maybe a typed array could be returned (e.g. Uint32Array) which would be more compact?

Reply:
Will think how to improve it to use typed arrays but for now all batch ids functions chain waits for simple array type.